compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -477,6 +529,66 @@ pub fn normalize_param_env_or_error<'tcx>(
     ty::ParamEnv::new(tcx.mk_clauses(&predicates))
 }
 
+// FIXME(-Zhigher-ranked-assumptions): this is a hack to walk around the fact that we don't support
+// placeholder assumptions right now.
+// We should remove this once we have proper support for implied bounds on binders.
+/// Deeply normalize the param env using the next solver.
+#[instrument(level = "debug", skip(tcx))]
+pub fn deeply_normalize_param_env_or_error<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    unnormalized_env: ty::ParamEnv<'tcx>,
+    cause: ObligationCause<'tcx>,
+) -> ty::ParamEnv<'tcx> {
+    let mut predicates: Vec<_> =
+        util::elaborate(tcx, unnormalized_env.caller_bounds().into_iter()).collect();
+
+    debug!("normalize_param_env_or_error: elaborated-predicates={:?}", predicates);
+
+    let elaborated_env = ty::ParamEnv::new(tcx.mk_clauses(&predicates));
+    if !elaborated_env.has_aliases() {
+        return elaborated_env;
+    }
+
+    let outlives_predicates: Vec<_> = predicates
+        .extract_if(.., |predicate| {
+            matches!(predicate.kind().skip_binder(), ty::ClauseKind::TypeOutlives(..))
+        })
+        .collect();
+
+    debug!(
+        "normalize_param_env_or_error: predicates=(non-outlives={:?}, outlives={:?})",
+        predicates, outlives_predicates
+    );
+    let Ok(non_outlives_predicates) =
+        do_deeply_normalize_predicates(tcx, cause.clone(), elaborated_env, predicates)
+    else {
+        // An unnormalized env is better than nothing.
+        debug!("normalize_param_env_or_error: errored resolving non-outlives predicates");
+        return elaborated_env;
+    };
+
+    debug!("normalize_param_env_or_error: non-outlives predicates={:?}", non_outlives_predicates);
+
+    // Not sure whether it is better to include the unnormalized TypeOutlives predicates
+    // here. I believe they should not matter, because we are ignoring TypeOutlives param-env
+    // predicates here anyway. Keeping them here anyway because it seems safer.
+    let outlives_env = non_outlives_predicates.iter().chain(&outlives_predicates).cloned();
+    let outlives_env = ty::ParamEnv::new(tcx.mk_clauses_from_iter(outlives_env));
+    let Ok(outlives_predicates) =
+        do_deeply_normalize_predicates(tcx, cause, outlives_env, outlives_predicates)
+    else {
+        // An unnormalized env is better than nothing.
+        debug!("normalize_param_env_or_error: errored resolving outlives predicates");
+        return elaborated_env;
+    };
```

Comment:
don't need to split outlives vs non-outlives predicate norm in the new solver. This only matters for the old one

Reply:
I think that way you also don't need a separate `do_deeply_normalize_predicates` function as you only call it once